### PR TITLE
Make clusteringress scalable

### PR DESCRIPTION
**Pull request by Miciah**
Add `/scale` and `/status` subresources to clusteringress, add fields to spec and status to facilitate scaling, and modify the controller to scale the router in accordance with the clusteringress.

This commit implements [NE-127](https://jira.coreos.com/browse/NE-127).

* `assets/defaults/cluster-ingress.yaml`: Set default replicas value to 1.
* `manifests/00-cluster-role.yaml`: Allow operator to update clusteringress status.
* `manifests/00-custom-resource-definition.yaml`: Add `/scale` and `/status` subresources.
* `pkg/apis/ingress/v1alpha1/clusteringress_types.go` (`ClusterIngressSpec`): Add replicas field.
(`ClusterIngressStatus`): Add replicas and selector fields.
* `pkg/manifests/bindata.go`: Regenerate.
* `pkg/manifests/manifests.go` (`RouterDeployment`): Copy replicas value from the clusteringress spec to the router deployment spec.
* `pkg/manifests/manifests_test.go` (`TestManifests`): Make sure deployment's replicas value gets set to the value from the clusteringress or to an appropriate default.
* `pkg/operator/controller/controller.go` (`ensureRouterForIngress`): Update the clusteringress's status, using `syncClusterIngressStatus`.
(`syncClusterIngressStatus`): New function that updates the given clusteringress's status with the current replicas and selector.
(`deploymentConfigChanged`): Check the deployment's replicas field.
* `test/e2e/operator_test.go` (`TestClusterIngressScale`): New test that checks the clusteringress's status and tries to scale the router up and down using the scale client.
(`getScaleClient`): New helper function for `TestClusterIngressScale`.

### CODE REVIEW

File: pkg/apis/ingress/v1alpha1/clusteringress_types.go [hunk 1]
```diff
@@ -75,8 +78,15 @@ type ClusterIngressHighAvailability struct {
...
-	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+
+	// Replicas is the actual number of observed router instances.
+	Replicas int32 `json:"replicas"`
+
+	// Selector is a label selector, in string format, for pods
+	// corresponding to the ClusterIngress. The number of matching pods
+	// should equal the value of replicas.
+	Selector string `json:"labelSelector"`
```
Comment (ironcladlou) [inline pkg/apis/ingress/v1alpha1/clusteringress_types.go]: Can you help me understand the rationale for this field?

Comment (Miciah) [reply pkg/apis/ingress/v1alpha1/clusteringress_types.go]: It is required by the HPA.  https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#scale-subresource

**Comment by Miciah** [comment]
Test "Create a release image in the release image stream" is failing:

    info: Included 46 referenced images into the payload
    error: unable to create a release: operator "cluster-osin-operator" failed to map images: image file "/tmp/release-image-0.0.1-2019-01-07-224105038230477/cluster-osin-operator/image-references" referenced image "cluster-authentication-operator" that is not part of the input images

/retest

File: pkg/apis/ingress/v1alpha1/clusteringress_types.go (hunk 1 above)
Comment (ironcladlou) [reply pkg/apis/ingress/v1alpha1/clusteringress_types.go]: Excellent, thanks

**Comment by ironcladlou** [comment]
@Miciah safe to remove the WIP/hold?

**Comment by Miciah** [comment]
@ironcladlou, if you're willing to trust `TestClusterIngressScale`, then yeah, but I have done no manual testing.

**Comment by ironcladlou** [comment]
@Miciah 
>@ironcladlou, if you're willing to trust TestClusterIngressScale, then yeah, but I have done no manual testing.

It'd be nice if we could trust it... let's start putting it to the test? 😁 Couldn't hurt to do a manual test in parallel and also see how the CLI behaves (any issues w/ the latter could be addressed in a followup)? I personally wouldn't block on either if the e2e works and if test environments are troublesome at the moment

**Comment by lihongan** [comment]
Two questions:
1. Does this mean we cannot using `oc -n openshift-ingress scale deployment/router-default` to control the replicas of router pod? Even we do that, the replicas of deployment will be synced to clusteringress, right?
2. (may not be related to this PR) but how to implement router sharding? and how to control the replicas of each router shards?

**Comment by lihongan** [comment]
@Miciah thanks for your quick reply:)

**Comment by Miciah** [comment]
/retest
Looks like https://github.com/openshift/release/issues/2070

**Comment by Miciah** [comment]
/retest
Looks like openshift/release#2070 again.

**Comment by Miciah** [comment]
2019/01/08 17:19:03 error: unable to signal to artifacts container to terminate in pod release-latest, triggering deletion: could not run remote command: unable to upgrade connection: container not found ("artifacts")

**Comment by Miciah** [comment]
level=fatal msg="waiting for openshift-console URL: context deadline exceeded"

/retest

**Comment by Miciah** [comment]
Same failure related to the console again.

/retest

**Comment by Miciah** [comment]
Rebased.

**Comment by Miciah** [comment]
error: unable to connect to image repository registry.svc.ci.openshift.org/ci-op-jx8ztfv8/stable@sha256:0bb5192165dd0eef1d4e1f23661f14c4e19a0ddba37bc73ad6d8db3a74cca176: Get https://registry.svc.ci.openshift.org/v2/: net/http: TLS handshake timeout

/retest

**Comment by Miciah** [comment]
2019/01/09 00:42:24 error: unable to signal to artifacts container to terminate in pod release-latest, triggering deletion: could not run remote command: unable to upgrade connection: container not found ("artifacts")

/retest

**Comment by Miciah** [comment]
error: could not run steps: test "release-latest" failed: pod release-latest was already deleted

/retest

**Comment by Miciah** [comment]
level=fatal msg="waiting for openshift-console URL: context deadline exceeded"

/retest

**Comment by Miciah** [comment]
level=fatal msg="waiting for openshift-console URL: context deadline exceeded"

/retest

**Comment by Miciah** [comment]
error: unable to parse image registry.svc.ci.openshift.org/ci-op-ilgn42p8/stable@sha256:ba8f8a00495858ce8b2d74106c2a674d2188598eebf282d80cd8bd75d820f663: cannot retrieve image configuration for manifest sha256:ba8f8a00495858ce8b2d74106c2a674d2188598eebf282d80cd8bd75d820f663: received unexpected HTTP status: 504 Gateway Time-out

/retest

**Comment by Miciah** [comment]
level=fatal msg="waiting for openshift-console URL: context deadline exceeded"

/retest

**Comment by Miciah** [comment]
level=error msg="\t* aws_vpc.new_vpc: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached."

**Comment by Miciah** [comment]
e2e-aws-operator passed.  Remaining tasks are (1) stop setting log level now that problems with the test are resolved and (2) refactor RBAC.

**Comment by Miciah** [comment]
Using this PR to test an IAM policy change that should resolve the `AccessDeniedException` errors in the operator's logs.

/test e2e-aws-operator

**Comment by Miciah** [comment]
`AccessDeniedException` exceptions are gone from the operator's logs.

**Comment by Miciah** [comment]
@openshift/sig-network-edge, tests are passing, and the PR is ready for review.

**Comment by Miciah** [comment]
Latest push changes `syncClusterIngressStatus` to use `deployment.Status.AvailableReplicas` instead of listing pods and deletes the now unneeded permission to list pods (thanks to @ironcladou for the suggestion!).

File: test/e2e/operator_test.go [hunk 2]
```diff
@@ -284,3 +292,165 @@ u3YLAbyW/lHhOCiZu2iAI8AbmXem9lW6Tr7p/97s0w==
...
 	return cl.Create(context.TODO(), secret), secret
 }
+
+func TestClusterIngressScale(t *testing.T) {
+	cl, ns, err := getClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	name := "default"
```
Comment (ironcladlou) [inline test/e2e/operator_test.go]: It just occurred to me... this is an e2e test running against an existing cluster which is the target of other tests (some of which may even be running in parallel). And the test is mutating the default ClusterIngress. Isn't that dangerous? If we need to test scaling through e2e, I wonder if we'd better create new ClusterIngresses to work against during tests like this?

Comment (Miciah) [reply test/e2e/operator_test.go]: This is a test in the cluster-ingress-operator repository, so it runs for the e2e-aws-operator job; doesn't that run in a separate cluster from the e2e-aws job? Also, this test is scaling the clusteringress up by 1 and then down by 1 to the original value, never below the original value, so it shouldn't break anything (or if it does, that might indicate a separate issue).

**Comment by Miciah** [comment]
/retest
Looks like openshift/release#2070 again.

**Pull request merged**